### PR TITLE
mrc-2582 Input time series table

### PR DESCRIPTION
**Pull request by EmmaLRussell**
## Description

Again, apologies for this bloated PR, I will try not to make a habit of it.. 

This branch adds the table associated with the input time series as shown in the [mockup](https://mrc-ide.myjetbrains.com/youtrack/api/files/112-1148?sign=MTYzMjcwMDgwMDAwMHwyNC0xN3wxMTItMTE0OHxyMVo5eUhScWFUUktYbFRDbXRXTVB6YWE0VEp2%0D%0ARDhoYm1NQVZ2bFZsUWgwDQo%0D%0A&updated=1629983996651). Because the input time series chart is defined as a generic chart, its associated table is also defined as part of this configuration. 

Note that some displayed column headers are still 'unfriendly' because we do not yet have label metadata for them, because they are not defined as filters in this chart. This will be resolved in https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2624 The affected columns are `area_name` and `time_period`. The last column, showing the selected plot type as header, is also awaiting friendly labels in this filter. 

The main changes are:

- adds table config to generic chart configuration. This is placed at the dataset level, since columns can differ between datasets. 
- refactors the existing Table component into two separate components:
   - `AreaIndicatorsTable` - which does all the data filtering and labelling data previously done in the Table component, which is geared to the non-generic HINT datasets, with known area filter, and a set of indicators separate from the filters (i.e. a different usage from the input time series). This passes processed data to:
   - `Table` - this is now a much simpler component, which just takes pre-filtered data and fields and renders them into a BTable. It passes through any scoped slot templates from the parent (this is used by `AreaIndicatorsTable` to show area hierarchy and indicator uncertainty ranges).
- adds a `GenericChartTable` component, which takes pre-filtered data along with table config and other generic chart configuration, and renders a `Table` component with labelled data and fields.
- updates `GenericChart` to render a `GenericChartTable` per dataset (there is only one for input time series) where configured. 
 
The table config consists of an array of `columns` where each column specifies:
-  `data`
   -  `columnId` - the column id in the dataset from which table values should be taken for this column
   -  `labelFilterId` - if non-null, the filter whose options should be used to find labels associated with column values, which will be displayed in the table instead of the raw values. e.g. for the age filter, this allows us to render '15-49' instead of 'Y015_049'
- `header`
  -  `type` - the options here are `filterLabel` or `selectedFilterOption`. if `filterLabel`, the label of the filter object specified by `filterId` will be used as the column header e.g. 'Age'. If `selectedFilterOption`, the label of the specified filter's selected option will be used as the column header. This allows us to set the header of the value column to the label of the selected plot type, e.g. 'HIV prevalence'. 
  - `filterId` - the filter from which to take filter label or selected filter option label. 

This configuration is subject to change once https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2624 is implemented, since we will no longer be using 'filter' metadata but 'column' metadata, so it will be appropriate to update the config schema at that point.

## Type of version change
_Delete as appropriate_

None

## Checklist

- [x] I have incremented version number, or version needs no increment
- [x] The build passed successfully, or failed because of ADR tests

### CODE REVIEW

**Review by mwoodbri — COMMENTED** [review]
Design and implementation both look very good to me. A couple of comments inline, and a couple of questions below. Only other observation with hindsight is that it might have made sense, if possible, to break this into two PRs: one to split existing table into two components, and another to add generic table on top.

1. This is a part of the app I haven't worked on, so this is definitely a question rather than an accusation, but should I always see a table below the bar chart? I am in some instances, but not here, and just wondering why:
Screenshot from 2021-09-27 10-57-58

2. There seems to be change (potentially a regression?) in how table values are displayed (before and after):
Screenshot from 2021-09-27 11-00-55

Screenshot from 2021-09-27 11-01-19

**Review by carbowizzy — CHANGES_REQUESTED** [review]
Looks good and works well. 

A big concern for me is the extra hyphen as seen in @mwoodbri screenshot above. I have also added some suggestion, you have not really changed some of these codes but since you have branched out a component from it, it would be nice to attend to them.

File: src/app/static/src/app/components/plots/table/Table.vue [hunk 1]
```diff
@@ -24,7 +24,7 @@
                 responsive="sm"
                 show-empty>
                 <template v-for="(_, slot) in $scopedSlots" v-slot:[slot]="props">
-                    <slot :name="slot" v-bind="props" />-
+                    <slot :name="slot" v-bind="props" />
```
Comment (EmmaLRussell) [inline src/app/static/src/app/components/plots/table/Table.vue:27]: This was the culprit responsible for the rogue additional '-' lines... :facepalm:

**Comment by EmmaLRussell** [comment]
> should I always see a table below the bar chart? 

@mwoodbri You should, yes. I am seeing the table for the Antarctica dataset with my ADR key, running locally. From your screenshot, I'm a bit suspicious of the area filter saying "MWI (unknown)" - that suggests something's gone a bit wrong with the area filter, which could be impacting the table too. Which dataset were you using, or were you using local files?

**Comment by EmmaLRussell** [comment]
> I think we could fetch data differently as well, header columns in different array and data different array. This would have reduced a lot of filtering/ data manipulation.

We're going to be changing the input time series response slightly to replace 'filters' with 'columns', though the structure will be essentially the same. As discussed earlier, I think a general refactor of how we handle metadata vs data, clarifying responsibilities for labelling, and source of truth for metadata etc would be something valuable to tackle after the workshops, and that could include agreeing on most tractable data structures for readable front end code.

File: src/app/static/src/app/components/genericChart/GenericChartTable.vue [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                return fields
+            },
+            labelledData() {
+                const filtersDict = this.filters.reduce((dict, filter) => ({...dict, [filter.id]: filter}), {} as Dict<Filter>);
+                const result = this.filteredData.map(row => {
+                    const friendlyRow = {...row};
+                    this.tableConfig.columns.filter(column => column.data.labelFilterId).forEach(column => {
+                        const filter = filtersDict[column.data.labelFilterId!];
+                        const friendlyValue = filter.options.find(option => option.id == row[column.data.columnId])?.label;
+                        friendlyRow[column.data.columnId] = friendlyValue;
```
[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply src/app/static/src/app/components/genericChart/GenericChartTable.vue:74]: Not sure about this one. This is already a really long line and I don't really want to make it even longer. I'll change it if you feel strongly about it, but I think readability impaired by line length wins out over condensing lines here.

**Comment by mwoodbri** [comment]
> > should I always see a table below the bar chart?
> 
> @mwoodbri You should, yes. I am seeing the table for the Antarctica dataset with my ADR key, running locally. From your screenshot, I'm a bit suspicious of the area filter saying "MWI (unknown)" - that suggests something's gone a bit wrong with the area filter, which could be impacting the table too. Which dataset were you using, or were you using local files?

Ok - sorry - Burundi but I was comparing this branch with deployed version, rather than with master running locally. Actually an artefact of mock model. Can be ignored.

File: src/app/static/src/tests/components/plots/table/table.test.ts [hunk 3]
```diff
@@ -73,450 +48,181 @@ const getWrapperFr = (customPropsData: any = {}) => {
...
+    };
+
     it('renders a table', () => {
         const wrapper = getWrapper();
         expect(wrapper.contains('div')).toBe(true);
         expect(wrapper.contains('table')).toBe(true);
         expect(wrapper.contains('br')).toBe(true);
     });
+
+    //TODO: test for rendering slots
```
[the remark this answers is not in the document]
Comment (mwoodbri) [reply src/app/static/src/tests/components/plots/table/table.test.ts]: Presumably referring to [this](https://github.com/mrc-ide/hint/pull/554/files#diff-573f702159974d05c90d5fd1eb84d452fec223b1b0e919022fe568431c351f15R161)?

**Review by mwoodbri — APPROVED** [review]

File: src/app/static/src/tests/components/plots/table/table.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply src/app/static/src/tests/components/plots/table/table.test.ts]: Yes!

**Review by Nicolas-Dolan — APPROVED** [review]

**Review by carbowizzy — APPROVED** [review]
LGTM

just conflicts to resolve

**Pull request merged**